core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,239 @@
+
+  private val listener = new SparkListener {
+    override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
+      val stageInfo = stageCompleted.stageInfo
+      // Remove internal data from a finished stage attempt.
+      cleanupSyncRequests(stageInfo.stageId, stageInfo.attemptNumber)
+      barrierEpochByStageIdAndAttempt.remove((stageInfo.stageId, stageInfo.attemptNumber))
+      cancelTimerTask(stageInfo.stageId, stageInfo.attemptNumber)
+    }
+  }
+
+  // Epoch counter for each barrier (stage, attempt).
+  private val barrierEpochByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), Int]
+
+  // Remember all the blocking global sync requests for each barrier (stage, attempt).
+  private val syncRequestsByStageIdAndAttempt =
+    new ConcurrentHashMap[(Int, Int), ArrayBuffer[RpcCallContext]]
+
+  // Remember all the TimerTasks for each barrier (stage, attempt).
+  private val timerTaskByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), TimerTask]
+
+  // Number of tasks for each stage.
+  private val numTasksByStage = new ConcurrentHashMap[Int, Int]
+
+  override def onStart(): Unit = {
+    super.onStart()
+    listenerBus.addToStatusQueue(listener)
+  }
+
+  /**
+   * Get the array of [[RpcCallContext]]s that correspond to a barrier sync request from a stage
+   * attempt.
+   */
+  private def getOrInitSyncRequests(
+      stageId: Int,
+      stageAttemptId: Int,
+      numTasks: Int): ArrayBuffer[RpcCallContext] = {
+    val requests = syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      new ArrayBuffer[RpcCallContext](numTasks))
+    if (requests == null) {
+      syncRequestsByStageIdAndAttempt.get((stageId, stageAttemptId))
+    } else {
+      requests
+    }
+  }
+
+  /**
+   * Clean up the array of [[RpcCallContext]]s that correspond to a barrier sync request from a
+   * stage attempt.
+   */
+  private def cleanupSyncRequests(stageId: Int, stageAttemptId: Int): Unit = {
+    val requests = syncRequestsByStageIdAndAttempt.remove((stageId, stageAttemptId))
+    if (requests != null) {
+      requests.clear()
+    }
+    logInfo(s"Removed all the pending barrier sync requests from Stage $stageId (Attempt " +
+      s"$stageAttemptId).")
+  }
+
+  /**
+   * Get the barrier epoch that correspond to a barrier sync request from a stage attempt.
+   */
+  private def getOrInitBarrierEpoch(stageId: Int, stageAttemptId: Int): Int = {
+    val defaultBarrierEpoch = 0
+    val barrierEpoch = barrierEpochByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      defaultBarrierEpoch)
+    if (barrierEpoch == null) {
+      defaultBarrierEpoch
+    } else {
+      barrierEpoch
+    }
+  }
+
+  /**
+   * Increase the barrier epoch that correspond to a barrier sync request from a stage attempt.
+   */
+  private def increaseBarrierEpoch(stageId: Int, stageAttemptId: Int): Unit = {
+    val barrierEpoch = barrierEpochByStageIdAndAttempt.get((stageId, stageAttemptId))
```

Comment:
This increment is not atomic. As commented [here](https://github.com/apache/spark/pull/21898#discussion_r206707616), do we need to use `AtomicLong` instead of `int` as `value` of `barrierEpochByStageIdAndAttempt`?